src/SparseMFModel.cpp

```diff
@@ -28,333 +62,459 @@ void SparseMFModel::initialize_weights(uint64_t users, uint64_t items, uint64_t
-          item_weights_lst.push_back(itemId);
         }
-        //std::cout << "UPDATE HERE " << std::endl;
-        item_weights_grad_map[itemId][k] += delta_item_w;
+        // XXX moving this after this inner loop
+        // gradient->users_weights_grad.push_back(std::make_pair(real_user_id,
+        // std::move(user_weights_grad)));
+
+        // update item latent factors
+        for (uint64_t k = 0; k < nfactors_; ++k) {
+          // std::cout << "k: " << k << std::endl;
+          FEATURE_TYPE delta_item_w =
+              learning_rate * (error * get_user_weights(user_from_0, k) -
+                               item_fact_reg_ * get_item_weights(itemId, k));
+          item_models[itemId].second[k] += delta_item_w;
+
+          if (item_weights_grad_map[itemId].size() == 0) {
+            item_weights_grad_map[itemId].resize(NUM_FACTORS);
+            item_weights_lst.push_back(itemId);
+          }
+          // std::cout << "UPDATE HERE " << std::endl;
+          item_weights_grad_map[itemId][k] += delta_item_w;
 #ifdef DEBUG
-        if (std::isnan(get_item_weights(itemId, k)) ||
-            std::isinf(get_item_weights(itemId, k))) {
-          std::cout << "error: " << error << std::endl;
-          std::cout << "rating: " << rating << std::endl;
-          std::cout << "pred: " << pred << std::endl;
-          std::cout << "delta_item_w: " << delta_item_w << std::endl;
-          std::cout << "user weight: " << get_user_weights(user_from_0, k) << std::endl;
-          std::cout << "item weight: " << get_item_weights(itemId, k) << std::endl;
-          std::cout << "learning_rate: " << learning_rate << std::endl;
-          throw std::runtime_error("nan in item weight");
-        }
+          if (std::isnan(get_item_weights(itemId, k)) ||
+              std::isinf(get_item_weights(itemId, k))) {
+            std::cout << "error: " << error << std::endl;
+            std::cout << "rating: " << rating << std::endl;
+            std::cout << "pred: " << pred << std::endl;
+            std::cout << "delta_item_w: " << delta_item_w << std::endl;
+            std::cout << "user weight: " << get_user_weights(user_from_0, k)
+                      << std::endl;
+            std::cout << "item weight: " << get_item_weights(itemId, k)
+                      << std::endl;
+            std::cout << "learning_rate: " << learning_rate << std::endl;
+            throw std::runtime_error("nan in item weight");
+          }
 #endif
+        }
+        // gradient->items_weights_grad.push_back(
+        //    std::make_pair(itemId, std::move(item_weights_grad)));
       }
-      //gradient->items_weights_grad.push_back(
-      //    std::make_pair(itemId, std::move(item_weights_grad)));
+      gradient->users_weights_grad.push_back(
+          std::make_pair(real_user_id, std::move(user_weights_grad)));
+      // std::cout
+      //  << "user weights size: " << gradient->users_weights_grad.size()
```

Comment:
Why do I have to tell you to fix these comments?

Reply:
Fixed